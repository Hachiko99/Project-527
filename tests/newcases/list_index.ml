type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

let list_index : list -> nat -> nat |>
  { [] => ( 0 => 0
          | 1 => 0 )
  | [2] => ( 0 => 0
           | 1 => 0
           | 2 => 0 )
  | [1; 2] => ( 0 => 0
              | 1 => 0
              | 2 => 1 )
  | [1] => ( 0 => 0
           | 1 => 0 )
  | [2; 1] => ( 2 => 0
              | 1 => 1 )
  | [3; 2; 1] => ( 3 => 0
                 | 2 => 1
                 | 1 => 2 )
  | [1; 2; 3; 4] => ( 1 => 0
                    | 2 => 1
                    | 3 => 2
                    | 4 => 3)
  } = ?
