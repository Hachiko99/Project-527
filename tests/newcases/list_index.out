(* Generated Tue Apr 19 12:38:27 2016 *)
let list_index : nat -> list -> nat =
  fun (n1:nat) ->
    let rec f2 (l1:list) : nat =
      match l1 with
        | Nil -> O
        | Cons (n2, l2) -> (match nat_equal n2 n1 with
                              | True -> O
                              | False -> (match l2 with
                                            | Nil -> O
                                            | Cons (n3, l3) -> S (f2 l2)))
    in
      f2
;;
